File: desktop/apps/desktop/src/lib/playQueue.js
// playQueue.js — the desktop app's "what plays next", as pure functions plus a
// tiny shared store. Playlists (Playlists.jsx) load it; "Play next" and "Add to
// queue" on any poster (AddToPlaylist.jsx) change it. No React, no IPC, so
// node --test checks it (test/play-queue.test.js).
//
// A queue is { items: [{ kind, id, title, entryId? }], pos, playlistId, shuffle, seed }.
// pos is the index of the item playing now; -1 means nothing has started.

export const EMPTY = Object.freeze({ items: [], pos: -1, playlistId: null, shuffle: false, seed: 0 })

const sameItem = (a, b) => !!a && !!b && a.id === b.id && (a.kind || 'movie') === (b.kind || 'movie')

/** A queue made from a playlist's play order, starting at `startIndex`. */
export function fromPlaylist(items, { startIndex = 0, playlistId = null, shuffle = false, seed = 0 } = {}) {
  const list = (Array.isArray(items) ? items : []).map((it) => ({ kind: it.kind === 'tv' ? 'tv' : 'movie', id: String(it.id), title: it.title || '', entryId: it.entryId || null, stream: it.stream || null, poster: it.poster || null, resumeSeconds: Number(it.resumeSeconds) || 0 }))
  const start = Math.max(0, Math.min(list.length - 1, Math.floor(Number(startIndex) || 0)))
  return { items: list, pos: list.length ? start - 1 : -1, playlistId, shuffle: !!shuffle, seed: Number(seed) || 0 }
}

/** The item that plays after the current one, or null. */
export function peekNext(q) {
  return (q && q.items[q.pos + 1]) || null
}

export function current(q) {
  return (q && q.pos >= 0 && q.items[q.pos]) || null
}

/** Move forward one. Returns the new queue (pos past the end means finished). */
export function advance(q) {
  if (!q || q.pos + 1 >= q.items.length) return { ...q, pos: q ? q.items.length : -1 }
  return { ...q, pos: q.pos + 1 }
}

export function back(q) {
  if (!q || q.pos <= 0) return q
  return { ...q, pos: q.pos - 1 }
}

/** Jump to an index (e.g. picking an item in the list). */
export function jump(q, index) {
  const i = Math.floor(Number(index))
  if (!q || !(i >= 0 && i < q.items.length)) return q
  return { ...q, pos: i }
}

/** "Play next": straight after what is playing now, ahead of the rest. */
export function playNext(q, items) {
  const add = (Array.isArray(items) ? items : [items]).filter(Boolean)
  const base = q || EMPTY
  const list = base.items.slice()
  list.splice(Math.max(0, base.pos + 1), 0, ...add)
  return { ...base, items: list }
}

/** "Add to queue": at the end. */
export function addToQueue(q, items) {
  const add = (Array.isArray(items) ? items : [items]).filter(Boolean)
  const base = q || EMPTY
  return { ...base, items: base.items.concat(add) }
}

export function remove(q, index) {
  if (!q || index < 0 || index >= q.items.length) return q
  const items = q.items.slice()
  items.splice(index, 1)
  const pos = index < q.pos ? q.pos - 1 : index === q.pos ? q.pos - 1 : q.pos
  return { ...q, items, pos }
}

/**
 * Something outside the queue was started (a poster clicked). If it is the next
 * queued item, the queue moves onto it; otherwise the queue carries on after it.
 */
export function startedOutside(q, item) {
  if (!q) return q
  for (let i = Math.max(0, q.pos + 1); i < q.items.length; i++) {
    if (sameItem(q.items[i], item)) return { ...q, pos: i }
  }
  return q
}

export function remaining(q) {
  return q ? Math.max(0, q.items.length - q.pos - 1) : 0
}

// --- one shared queue for the whole renderer ---------------------------------
let state = EMPTY
const listeners = new Set()

export function getQueue() {
  return state
}

export function setQueue(next) {
  state = next || EMPTY
  for (const fn of listeners) {
    try {
      fn(state)
    } catch {
      /* one broken listener must not stop the rest */
    }
  }
  return state
}

export function subscribe(fn) {
  listeners.add(fn)
  return () => listeners.delete(fn)
}

File: apps/core/app/src/main/java/com/beeboentertainment/movie/core/PlayQueue.kt
package com.beeboentertainment.movie.core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * "What plays next": a playlist in order or shuffled, plus anything added with Play next or
 * Add to queue. Pure data and rules, so the unit tests cover all of it; [PlayQueueHolder] is the
 * one shared instance the screens, the player and PlaybackService read.
 *
 * The same model as the website's and the desktop app's queue:
 *   - [pos] is the index of the item playing now; -1 before the first one starts.
 *   - When something outside the queue is played, the queue waits and carries on after it.
 *   - When the next queued item starts (from the player's ⏭, the up-next card, a car, a Cast
 *     receiver's controls), [onItemStarted] moves onto it.
 */
data class QueueItem(
    val kind: String,
    val id: String,
    val title: String = "",
    val stream: String? = null,
    val poster: String? = null,
    val showKey: String? = null,
    val entryId: String? = null,
    val resumeSeconds: Double = 0.0
) {
    fun sameAs(kind: String?, id: String?): Boolean = id == this.id && normalKind(kind) == normalKind(this.kind)

    companion object {
        fun normalKind(kind: String?): String = if (kind == "tv" || kind == "episode") "tv" else "movie"
    }
}

data class PlayQueue(
    val items: List<QueueItem> = emptyList(),
    val pos: Int = -1,
    val playlistId: String? = null,
    val playlistName: String? = null,
    val shuffle: Boolean = false,
    val seed: Long = 0
) {
    val current: QueueItem? get() = items.getOrNull(pos)
    val next: QueueItem? get() = items.getOrNull(pos + 1)
    val previous: QueueItem? get() = if (pos > 0) items.getOrNull(pos - 1) else null
    val remaining: Int get() = (items.size - pos - 1).coerceAtLeast(0)
    val isEmpty: Boolean get() = items.isEmpty()

    /** Move forward one; past the end means the queue has finished. */
    fun advance(): PlayQueue = copy(pos = (pos + 1).coerceAtMost(items.size))

    fun back(): PlayQueue = if (pos <= 0) this else copy(pos = pos - 1)

    fun jump(index: Int): PlayQueue = if (index in items.indices) copy(pos = index) else this

    /** Straight after the item playing now, ahead of everything else. */
    fun playNext(add: List<QueueItem>): PlayQueue {
        if (add.isEmpty()) return this
        val at = (pos + 1).coerceIn(0, items.size)
        return copy(items = items.subList(0, at) + add + items.subList(at, items.size))
    }

    /** At the end. */
    fun addToQueue(add: List<QueueItem>): PlayQueue = if (add.isEmpty()) this else copy(items = items + add)

    fun removeAt(index: Int): PlayQueue {
        if (index !in items.indices) return this
        val newPos = if (index <= pos) pos - 1 else pos
        return copy(items = items.filterIndexed { i, _ -> i != index }, pos = newPos)
    }

    /**
     * Something started playing. If it is a queued item still ahead, the queue moves onto it;
     * if it is the current one nothing changes; anything else leaves the queue waiting.
     */
    fun onItemStarted(kind: String?, id: String?): PlayQueue {
        if (id.isNullOrBlank()) return this
        if (current?.sameAs(kind, id) == true) return this
        for (i in (pos + 1).coerceAtLeast(0) until items.size) {
            if (items[i].sameAs(kind, id)) return copy(pos = i)
        }
        return this
    }

    /**
     * The next item for the player's transport when [id] is playing, or null when the queue has
     * nothing to say (the server's own up next applies):
     *   - [id] is the current queued item: the queue's next;
     *   - nothing from the queue has started yet (titles lined up with Play next / Add to queue):
     *     the first of them, after whatever is playing;
     *   - otherwise (an old, abandoned queue while something unrelated plays): null, so a queue
     *     from yesterday never hijacks today's film.
     */
    fun nextAfter(kind: String?, id: String?): QueueItem? {
        if (id.isNullOrBlank() || isEmpty) return null
        return if (current?.sameAs(kind, id) == true || pos == -1) next else null
    }

    /** The queue's previous item, only while [id] is the current queued item. */
    fun previousBefore(kind: String?, id: String?): QueueItem? =
        if (current?.sameAs(kind, id) == true) previous else null

    companion object {
        /** A queue from a playlist's play order, ready to start at [startIndex]. */
        fun fromPlaylist(
            items: List<QueueItem>,
            startIndex: Int = 0,
            playlistId: String? = null,
            playlistName: String? = null,
            shuffle: Boolean = false,
            seed: Long = 0
        ): PlayQueue {
            if (items.isEmpty()) return PlayQueue(playlistId = playlistId, playlistName = playlistName)
            val start = startIndex.coerceIn(0, items.size - 1)
            return PlayQueue(items, start - 1, playlistId, playlistName, shuffle, seed)
        }
    }
}

/** The one queue for the whole app, for as long as the process lives. */
object PlayQueueHolder {
    private val _queue = MutableStateFlow(PlayQueue())
    val queue: StateFlow<PlayQueue> = _queue.asStateFlow()
    val current: PlayQueue get() = _queue.value

    fun set(q: PlayQueue) { _queue.value = q }
    fun clear() { _queue.value = PlayQueue() }
    fun update(fn: (PlayQueue) -> PlayQueue) { _queue.value = fn(_queue.value) }

    /**
     * Called by PlaybackService whenever an item starts. Returns true when the queue moved onto
     * it, so the caller can report playlist progress for exactly those starts.
     */
    fun onItemStarted(kind: String?, id: String?): Boolean {
        val before = _queue.value
        val after = before.onItemStarted(kind, id)
        _queue.value = after
        return after.pos != before.pos
    }
}
